src/manifest/assertions/BMFFHashAssertion.ts

```diff
@@ -129,64 +267,174 @@ export class BMFFHashAssertion extends Assertion {
+        const chunks: Uint8Array[] = [];
+        const mdatBox = asset.getBoxByPath('/mdat');
+
+        if (!mdatBox) {
+            throw new ValidationError(ValidationStatusCode.AssertionBMFFHashMismatch, this.sourceBox, 'mdat not found');
+        }
+
+        // Per spec 15.12.2: Handle fixed and variable size blocks
+        if (tree.fixedBlockSize) {
+            let offset = mdatBox.payloadOffset;
+            for (let i = 0; i < tree.count; i++) {
+                chunks.push(await asset.getDataRange(offset, tree.fixedBlockSize));
+                offset += tree.fixedBlockSize;
+            }
+        } else if (tree.variableBlockSizes) {
+            let offset = mdatBox.payloadOffset;
+            for (const size of tree.variableBlockSizes) {
+                chunks.push(await asset.getDataRange(offset, size));
+                offset += size;
             }
         }
 
-        return box;
+        return chunks;
+    }
+
+    /**
+     * Updates the assertion with hash and exclusion data from the given asset
+     * @param asset - The asset to generate the hash from
+     * @throws {Error} If the asset is not a BMFF asset or if the algorithm is not set
+     */
+    public async updateWithAsset(asset: Asset): Promise<void> {
+        if (!this.algorithm) {
+            throw new Error('Assertion has no algorithm');
+        }
+
+        if (!(asset instanceof BMFF)) {
+            throw new Error('Asset must be a BMFF asset');
+        }
+
+        this.hash = await this.hashBMFFWithExclusions(asset);
+    }
+
+    /**
+     * Creates a new BMFFHashAssertion with the given algorithm and version 2
+     * @param name - The name of the assertion
+     * @param algorithm - The hash algorithm to use
+     * @returns {BMFFHashAssertion} The new BMFFHashAssertion
+     */
+    public static createV2(name: string, algorithm: HashAlgorithm) {
+        const bmffHashAssertion = new BMFFHashAssertion(2);
+        bmffHashAssertion.name = name;
+        bmffHashAssertion.algorithm = algorithm;
+        bmffHashAssertion.hash = new Uint8Array(Crypto.getDigestLength(algorithm));
+
+        bmffHashAssertion.exclusions = [
+            {
+                xpath: '/uuid',
+                data: [
+                    {
+                        offset: 8,
+                        value: new Uint8Array(BMFF.c2paBoxUserType),
+                    },
+                ],
+            },
+            {
+                xpath: '/ftyp',
+            },
+            {
+                xpath: '/mfra',
+            },
+        ];
+
+        return bmffHashAssertion;
+    }
+
+    /**
+     * Creates a new BMFFHashAssertion with the given algorithm and version 3
+     * @param name - The name of the assertion
+     * @param algorithm - The hash algorithm to use
+     * @returns {BMFFHashAssertion} The new BMFFHashAssertion
+     */
+    public static createV3(name: string, algorithm: HashAlgorithm) {
```

Comment:
The `createV2` and `createV3` methods share the majority of their code, could be streamlined.